numpy/distutils/ccompiler_opt.py

```diff
@@ -0,0 +1,1449 @@
+        )
+
+        # '_default_dispatch' : default dispatched CPU features, if they are supported by C compiler
+        self._default_dispatch = dict(
+            x86 = [
+                "SSE3",  "SSSE3", "SSE41", "POPCNT", "SSE42", "AVX",
+                "FP16C", "XOP", "FMA4", "FMA3", "AVX2", "AVX512_COMMON",
+                "AVX512_KNL", "AVX512_KNM", "AVX512_SKX", "AVX512_CNL", "AVX512_CEL", "AVX512_ICL"
+            ],
+            x86_64  = [
+                "SSSE3", "SSE41", "POPCNT", "SSE42", "AVX",
+                "FP16C", "XOP", "FMA4", "FMA3", "AVX2", "AVX512_COMMON",
+                "AVX512_KNL", "AVX512_KNM", "AVX512_SKX", "AVX512_CNL", "AVX512_CEL", "AVX512_ICL"
+            ],
+            ppc64   = ["VSX2", "VSX3"],
+            ppc64le = ["VSX3"],
+            arm     = ["NEON", "NEON_FP16", "NEON_VFPV4", "ASIMD", "ASIMDHP", "ASIMDDP", "ASIMDFHM"],
+            aarch64 = ["ASIMDHP", "ASIMDDP", "ASIMDFHM"],
+        )
+
+        """
+        The following nested dictionaries are used to configure CPU features,
+        the primary key is represented as a feature name or group name that gathers several features.
+        The other subkeys are represented as follows:
+
+        'implies' : str or list, optional
+            Set a feature or a list of features to be implied
+
+        'implies_flags' : bool, optional
+            By default, all flags of implied features are combined with the parent feature flags
+            but in some compilers, we need to disable this behavior.
+
+        'alias' : str or list of str, optional
+            It is only used to add extra definitions into the CPU dispatch header.
+
+        'interest' : int, required
+        'headers' : str or list, optional
+        'flags' : str or list, optional
+        """
+
+        # '_cpu_features' contains all known CPU features
+        self._all_features = self._dict_features(
+            # X86
+            SSE    = dict(headers="xmmintrin.h", interest=1),
+            SSE2   = dict(implies="SSE",    headers="emmintrin.h", interest=2),
+            SSE3   = dict(implies="SSE2",   headers="pmmintrin.h", interest=3),
+            SSSE3  = dict(implies="SSE3",   headers="tmmintrin.h", interest=4),
+            SSE41  = dict(implies="SSSE3",  headers="smmintrin.h", interest=5),
+            POPCNT = dict(implies="SSE41",  headers="popcntintrin.h", interest=6),
+            SSE42  = dict(implies="POPCNT", interest=7),
+            AVX    = dict(implies="SSE42",  headers="immintrin.h", interest=8),
+            FP16C  = dict(implies="AVX",    interest=9),
+            XOP    = dict(implies="AVX",    headers="x86intrin.h", interest=10),
+            FMA4   = dict(implies="AVX",    headers="x86intrin.h", interest=11),
+            FMA3   = dict(implies="AVX",    interest=12),
+            AVX2   = dict(implies=["AVX", "FP16C", "FMA3"], interest=13),
+            AVX512_COMMON = dict(implies="AVX2", alias=["AVX512F", "AVX512CD"], interest=20),
```

Comment:
Is it safe to club AVX-512F and AVX-512CD into one? The Intel SDM does provide different feature flag bits for these two features. It might be safer to keep them separate?

Reply:
Yes, it's totally safe, all CPUs that support `AVX512F ` also support `AVX512CD` and we do check both features availability in runtime via CPUID.
https://github.com/numpy/numpy/blob/5ecdabfa03e0a01639eda10adc5cdd59cc03df88/numpy/core/src/common/npy_cpu.c#L298-L306

Reply:
That may be so for current CPU's but may not hold for future generations (the fact that they are separate means they might be segmented sometime in the future). And moreover, this assumption may not hold for VM's which might support just AVX-512F.

Reply:
The best guess they separated just because conflict detection never been part of AVX2 and AMD may not support it.

> That may be so for current CPU's but may not hold for future generations

++ Alright let's play it safe with `AVX512F` and separate them not gonna lose anything.

> this assumption may not hold for VM's which might support just AVX-512F

Usually, VMs emulate CPUs not certain features